### ISSUE TITLE
Ability to unit test Go chaincode

### ISSUE THREAD

**Issue by srderson** [hyperledger-archives/fabric#2119]
## Description

Unit testing chaincode is difficult today as there's no standard way to mock the shim.ChaincodeStub object. This is needed by anyone writing a chaincode. This needs code (possibly there's an existing Go framework that could be used) in addition to documentation.

**Comment by brad-gorman** [comment]
I'm having a go at doing this. I'm making an interface that will let you replace the shim (core/chaincode/shim/chaincode.go) in the unit test environment with a much simpler struct.

I'm going to have to figure out what the security context and event are used for as part of this.

**Comment by srderson** [comment]
Thanks @brad-gorman! Will assign to you. @JonathanLevi or @adecaro may be good contacts for security questions.

**Comment by srderson** [comment]
@christo4ferris could you add @brad-gorman to the Hyplerledger organization so I can assign this issue to him?

**Comment by brad-gorman** [comment]
I've had to change the method signatures where *shim.ChaincodeStub was involved to my new interface shim.ChaincodeStubInterface. This means a lot of files need updating, including any chaincode people have.

I've got a simple chaincode_example02 unit test working though.

Here's the commit on my branch. Feedback is welcome.
https://github.com/brad-gorman/fabric/commit/99c2c75ee31c064188b19c9dffd57f0a473c62b0

**Comment by brad-gorman** [comment]
I now have ChaincodeStub's RangeQueryState(), InvokeChaincode() and QueryChaincode() working with unittests for examples 2, 3 and 4.
https://github.com/brad-gorman/fabric/commit/7f40fd3fc564bdfddaf57b4e13964c49d3212bb7

**Comment by brad-gorman** [comment]
I'm stopping work on this until the maintainers make a call on where the interface should be. Personally I feel it should an an interface that the user-written chaincode uses and is tested against - not a struct - even if that means a disruption to chaincode developers.

**Comment by ghaskins** [comment]
@brad-gorman I didn't mean to discourage you.   I was just curious.  I don't yet understand the nuance here to have a strong opinion.  To be clear, I am not against this becoming an interface at all.  I was merely questioning whether it had to be done in a way that was disruptive to existing code.  Therefore, let me rephrase my question:

Is there a way that we can get the desirable property of abstracting this into an interface (rather than a struct) that doesn't require chaincode applications to modify their code?  (recompilation requirement is fine, of course).  Even if there isn't, it's not a show-stopper.  I agree with the comments that it's better to get this right now rather than be stuck with a suboptimal design for the foreseeable future.  At the same time, if we can avoid breaking virtually every app out there and still retain the design improvement, that is even better.

**Comment by brad-gorman** [comment]
The only thing I can think that might work is really, really ugly. We'd need to add a constructor to ChaincodeStub, that populates a new MockStub member. Then every function that ChaincodeStub implements would be modified to look like this:

> func (stub *ChaincodeStub) GetState(key string) ([]byte, error) {
>   // if we have a MockStub perform the unit test equivalent of the operation
>   if stub.mock != nil {
>       return stub.mock.GetState(key)
>   }
>   // perform the real operation
>   return handler.handleGetState(key, stub.UUID)
> }

There is a danger that some new functions may need to be added to the ChaincodeStub to handle unit testing type behaviour. These could then be abused by users during normal operations, outside of unit testing.

I'm not sure why users needing to change their code is an issue. The project is very early on, and I've seen massive architectural changes are being discussed. This is really a simple search and replace. Existing, deployed chaincode will not stop functioning either as it is a standalone binary.

**Comment by brad-gorman** [comment]
I've got this changeset in Gerrit now: https://gerrit.hyperledger.org/r/#/c/365/

Jenkins fails due to outside things like hyperledger/chaintool and hyperledger/fabric-test-resources which have not been updated yet.
